src/main/native/pairhmm/JavaData.h

```diff
@@ -3,25 +3,45 @@
 
 #include <vector>
 #include <debug.h>
+#include <exception>
+#include <string>
 #include "pairhmm_common.h"
 
+class JavaException : std::exception {
+  public:
+  const char* classPath;
+  const char* message;
+
+  JavaException(const char* classPath, const char* message)
+    : classPath(classPath),
+      message(message)
+  { }
+};
+
 class JavaData {
  public:
+  JavaData(JNIEnv *env) : m_env(env) {
+  
+  }
+
+  virtual ~JavaData() {
+    releaseData();
+  }
 
   // cache field ids
-  static void init(JNIEnv *env, jclass readDataHolder, jclass haplotypeDataHolder) {
```

Comment:
Not sure if this conforms to jni practices and  guarantees thread safety?

Reply:
I don't really know what part of the changes you're referring to. Regardless, objects of this class aren't being shared across threads (or separate JNI functions). The lifetime of the `JNIEnv` reference is the same as it's been before.

Reply:
Mostly the JNIEnv getting used as first parameter, but this is not JNI interface, so seems ok

Reply:
Can keep init as static